Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp

```diff
@@ -0,0 +1,173 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "FileSystemWritableFileStreamSink.h"
+
+#include "FileSystemFileHandle.h"
+#include "JSBlob.h"
+#include <JavaScriptCore/ArrayBufferView.h>
+#include <JavaScriptCore/JSArrayBuffer.h>
+#include <JavaScriptCore/JSArrayBufferView.h>
+#include <JavaScriptCore/JSString.h>
+
+namespace WebCore {
+
+static FileSystemWritableFileStream::WriteParams writeParamsFromChunk(FileSystemWritableFileStream::ChunkType&& chunk)
+{
+    FileSystemWritableFileStream::WriteParams result;
+    result.type = FileSystemWritableFileStream::WriteCommandType::Write;
+    WTF::switchOn(WTFMove(chunk), [&](FileSystemWritableFileStream::WriteParams&& params) {
+        result = WTFMove(params);
+    }, [&](RefPtr<JSC::ArrayBufferView>&& data) {
+        result.data = WTFMove(data);
+    }, [&](RefPtr<JSC::ArrayBuffer>&& data) {
+        result.data = WTFMove(data);
+    }, [&](RefPtr<Blob>&& data) {
+        result.data = WTFMove(data);
+    }, [&](String&& data) {
+        result.data = WTFMove(data);
+    });
+
+    return result;
+}
+
+static void fetchDataBytesForWrite(const FileSystemWritableFileStream::DataVariant& data, CompletionHandler<void(bool, std::span<const uint8_t>)>&& completionHandler)
+{
+    WTF::switchOn(data, [&](const RefPtr<JSC::ArrayBufferView>& bufferView) {
+        if (!bufferView || bufferView->isDetached())
+            return completionHandler(false, { });
+
+        RefPtr buffer = bufferView->possiblySharedBuffer();
+        if (!buffer)
+            return completionHandler(false, { });
+
+        completionHandler(true, buffer->span());
+    }, [&](const RefPtr<JSC::ArrayBuffer>& buffer) {
+        if (!buffer || buffer->isDetached())
+            return completionHandler(false, { });
+
+        completionHandler(true, buffer->span());
+    }, [&](const RefPtr<Blob>& blob) {
+        if (!blob)
+            return completionHandler(false, { });
+
+        blob->getArrayBuffer([completionHandler = WTFMove(completionHandler)](auto&& result) mutable {
+            if (result.hasException())
+                return completionHandler(false, { });
+
+            Ref buffer = result.releaseReturnValue();
+            if (buffer->isDetached())
```

Comment:
Can it be a case?
Maybe we can just ASSERT?

Reply:
I am not sure; will see if I could add a test.